<div id="static-content">
<div class="docbook-page">

  <div style="clear:" class="box">


<div class="heading2">FlyMine version 17.0 - release notes (12th June 2009)</div>
<div class="body">
<p style="font-size: +1.1em;">
In Release 17.0 we have updated data from FlyBase to release FB2009_05 and many sources are updated to the latest versions.</p>

<h3>Data</h3>
<p>Data from FlyBase is updated to the FB2009_05 release and many other sources have been updated to the latest versions, specific changes are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>

<ul>
 <li>NEW - Added natural_transposable_element features from FlyBase chado.</li>
 <li>FIX - Fixed bug that caused some proteins from FlyBase not to be loaded.</li>
 <li>UPDATE - Create synonyms for FlyBase proteins discarded because they have an identical sequence to a protein already stored in the database.</li>
 <li>UPDATE - Added seven new FlyAtlas tissues.</li>
 <li>FIX - Added datasets and long names to FlyBase interactions.</li>
 <li>FIX - Added datasets to protein domains from InterPro.</li>
 <li>FIX - Fixed bug that prevented the loading of self-self interactions from IntAct and BioGRID.</li>
</ul>

<h3>User interface</h3>
<p>
<ul>
 <li>UPDATE - Added symbol to identifiers to show in HTML title, eg changed 'Report page: FBgn0000606' to 'Gene report for eve (FBgn0000606)'</li>
 <li>NEW - Added link from list analysis and report pages to modMine.</li>
 <li>FIX - Improved performance of templates page in MyMine.</li>
 <li>FIX - Fixed bug encountered when trying to access the list web services.</li>
 <li>UPDATE - Changed the orientation of the FlyAtlas widget to increase readability.</li>
 <li>UPDATE - Improved error messages generated by the web services.</li>
 <li>UPDATE - Improved messages given when arriving to FlyMine from another website.</li>
 <li>FIX - Fixed link to QueryBuilder on home page.</li>
 <li>UPDATE - GOTerm now autocompletes in the QueryBuilder.</li>
 <li>FIX - Fixed bug in bulk download query for Anopheles homologues.</li>
</ul>
</p>


</div>

<div class="heading2">FlyMine version 16.0 - release notes (8th May 2009)</div>
<div class="body">
<p style="font-size: +1.1em;">
In Release 16.0 we have updated data from FlyBase to release FB2009_04 and many sources are updated to the latest versions, the way we include proteins has been redesigned so that all isoforms from UniProt are loaded.  We have introduced a significant new feature to querying ('outer joins') and improved general query and export performance.</p>

<h3>Data</h3>
<p>Data from FlyBase is updated to the FB2009_04 release and many other sources have been updated to the latest versions, specific changes are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>

<ul>
    <li>UPDATE - We now include all isoforms of proteins from UniProt.</li>   
    <li>UPDATE - Gene Ontology (GO) data assigned to genes has been re-modelled to now include relationship types between terms.  Genes no longer have an allGoAnnotation collection to model transitive closure of the ontology, instead use the GoAnnotation.relations.parents collection.</li>
    <li>NEW - We now load deficiencies/deletions and aberrations from FlyBase</li>
    <li>FIX - TransposableElementInsertionSite features now have orientations, these are displayed in GBrowse.</li>
    <li>FIX - Rat genes now have synonyms, organisms and datasets correctly attached </li>
    <li>FIX - Drosophila mRNA now have synonyms and datasets correctly attached</li>
    <li>FIX - ESTs now have locations</li>
    <li>FIX - Dmel proteins from UniProt are correctly attached to genes.</li>
    <li>FIX - All genes and proteins now have a datasource.</li>
    <li>UPDATE - Cytological location data now available</li>
    <li>UPDATE - Proteins and transcripts are now linked</li>
    <li>NEW - Predicted miRNA targets</li>
    <li>NEW - Drosophila development ontology</li>
    <li>NEW - Synonyms now have datasets, so you can tell who is using which identifier.</li>
    <li>UPDATE - Anopheles proteins now loaded for correct strain.</li>
    <li>UPDATE - <i>Apis mellifera</i> genome removed.</li>
</ul>

<h3>Query system</h3>
<p>In a major update to the query system it is now possible to choose whether columns include information that must be present to produce a row in the results or should just be shown if a value is present.  This is best illustrated with an example.  Consider an organism that has a thousand genes but only 900 are known to code for a protein.  In FlyMine there would be 1000 genes and 900 of them would be related to a protein.  You can now query for genes and related proteins in two ways:</p>
<ol>
  <li>'inner join' - You create a query with two columns, the first for the gene identifier, the second with the protein identifier.  Only 900 rows are return, those that don't have a protein aren't included.  This is how all queries worked before this release.</li>
  <li>'outer join' - You create the same query but specify the gene to protein relationship as an 'outer join'.  This will return all 1000 genes in the first column and have empty cells in the second column for the 100 genes that don't have a protein.</li>
</ol>

<h3>Organise your lists</h3>
<p>You can now apply arbitrary tags to your saved lists both on the list analysis page and in MyMine.  Tags can be any label you like to help you categorise your saved information, one list can have multiple tags applied.  You can filter lists by tag from the Lists view page.</p>
  
<h3>Export</h3>
<p>Improvements to the internals of InterMine have made exporting large datasets significantly faster - you should find download speeds around an order of magnitude better.</p>

<h3>User interface</h3>
<p>
<ul>
  <li>NEW - web service API is available.</li>
  <li>FIX - Creating a list after converting between orthologues now works.</li>
  <li>NEW - link to ArrayExpress atlas is available on certain report pages.</li>
  <li>NEW - link to FlyBase appears on list analysis pages for alleles and insertions.</li>
  <li>FIX - Foreign characters display properly</li>
  <li>FIX - Chromosome distribution widget now excludes rubbish chromosomes from its calculations.</li>
  <li>FIX - 404 error page improved.</li>
  <li>FIX - List information on the report page is now at the top, for increased visibility</li>
  <li>FIX - Export display issues fixed for Internet Explorer.</li>
  <li>FIX - Fixed some slow algorithms displaying results data.</li>
  <li>FIX - Fixed error in "Add Column" function where there are no extra columns to add.</li>
  <li>FIX - Fixed "View all" button on Sequence residues.</li>
  <li>UPDATE - Export now has column headings at the top.</li>
  <li>FIX - Performance of widget queries improved.</li>
  <li>FIX - Wildcards in queries are exported correctly.</li>
  <li>UPDATE - Collection sizes are updated using AJAX on report pages.</li>
  <li>UPDATE - UTRs now present in GBrowse</li>
</ul>
</p>
</div>
<div class="heading2">FlyMine version 15.0 - release notes (15th September 2008)</div>
<div class="body">
<p style="font-size: +1.1em;">
In Release 15.0 we have updated data from FlyBase to release FB2008_08 and many sources are updated to the latest versions.</p>

<h3>Data</h3>
<p>Data from FlyBase is updated to the FB2008_08 release and many other sources have been updated to the latest versions, specific changes are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>

<ul>
    <li>NEW - FlyMine now loads all types interactions from the BioGRID for D. melanogaster, C. elegans and S. cerevisiae. Currently, no distinction is made between protein and genetic interactions. All of them (the ones from IntAct as well as the ones from the BioGRID) are now called 'Interaction'. The Interaction Term allows to search for a particular interaction, eg phenotypic enhancement, two hybrid.</li>
    <li>NEW - Affymetrix Probesets from the GeneChip D. melanogaster Genome 1.0 and 2.0 Arrays now have locations and mapped genes, obtained from the Ensembl database. FlyMine only loads probesets that match to at least one gene. More information on the Ensembl probeset mappings, can be found at <a href="http://www.ensembl.org/Homo_sapiens/helpview?kw=microarray;ref=http%3A%2F%2Fwww.ensembl.org%2FMus_musculus%2Findex.html/" target="_new">Ensembl help</a> </li>
    <li>NEW - Loaded all aberrations from FlyBase including Drosdel deletions and they are now called ChromosomeStructureVariation. Each feature type is represented by a seperate sub-class: ChromosomalDeletion, ChromosomalDuplication, ChromosomalInversion, and ChromosomalTranslocation.</li>
    <li>FIXED - The Ensembl id is correctly set for human genes.</li>
    <li>NEW - Gene.flybaseFeatureType has been renamed to Gene.featureType.</li>
    <li>FIXED - all MRNAs now have Translations.</li>
    <li>FIXED - UTRs now have chromosome locations and appear in GBrowse.</li>
</ul>

<h3>Export</h3>
<p>In this release, we've extended and improved the export system.  We've optomised and configured the export process which has resulted in marked speed improvements.  Also, we've added new functionality. </p>


<h3>User interface</h3>
<p>
<ul>
  <li>NEW - you can now have a dash in a list name</li>
  <li>NEW - links on the allele list page directly to FlyBase</li>
  <li>NEW - links on the allele list page to convert list to genes</li>
  <li>FIXED - improved error handling throughout the site, including the QueryBuilder, template and query import pages</li>
  <li>FIXED - the quick search now handles special characters correctly so you can search for any identifier, eg. </li>
  <li>FIXED - selected records in your list stay highlighted even if you navigate away and then back to the list analysis page</li>
  <li>FIXED - template form handles wildcards correctly</li>
</ul>
</p>
</div>

<div class="heading2">FlyMine version 14.0 - release notes (14th July 2008)</div>
<div class="body">
<p style="font-size: +1.1em;">
In Release 14.0 we have updated data from FlyBase to release FB2008_06 and many sources are updated to the latest versions.  We have added an autocompletion feature for many constraints in template queries and QueryBuilder.  This means that as you type, for example a GO term or protein domain name, suggested matches will be provided.</p>

<h3>Data</h3>
<p>Data from FlyBase is updated to the FB2008_06 release and many other sources have been updated to the latest versions, specific changes are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>
<p>
<ul>
  <li>NEW - <i>Drosophila</i> genes now have a field 'flyBaseFeatureType' field for the gene type e.g. protein_coding, miRNA_gene.</li>
  <li>NEW - <i>D. melanogaster</i> microRNATargets now have chromosome locations and appear in GBrowse.</li>
  <li>FIX - <i>C. elegans</i> genes and proteins are now correctly linked to one another.</li>
  <li>NEW - CRM and binding sites from REDfly now have an evidence field which contains information about the experiment.</li>
  <li>FIX - More probesets from AffyMetrix are correctly mapped to genes, including those with predicted transcripts</li>
  <li>NEW - A new tissue, the salivary gland, has been added to the FlyAtlas dataset.</li>
</ul>
</p>

<h3>Data integration</h3>
<p>In this release we further improved the data integration system to update old identifiers for <i>Drosophila</i> genes that are provided by many data sources to match a current FlyBase identifier.  We now include only current FlyBase genes that have a genome location (see 13.0 release notes for more details).</p>

<h3>Data sources</h3>
<p>Most data, including all genes and proteins, have been assigned a dataset.  A dataset includes the source of the data, dataset name and URL, allowing you to quickly and easily find the origin of the data.</p>

<h3>Autocompletion</h3>
<p>Many name fields in template queries and the QueryBuilder now have type-ahead autocompletion to assist in selecting valid terms.  As you start to type, possible matches are fetched from FlyMine; the text you have typed can match anywhere within the terms and multiple words can be matched.  This is particularly useful for GO terms or protein domain names.</p>

<h3>Removing list entries</h3>
<p>It is now possible to remove entries from one of your lists.  On the List Analysis page you can now select entries in the list and use the new 'Edit' menu to delete them from your list or add them to another existing list.</p>

<h3>User interface</h3>
<p>
<ul>
  <li>NEW - added linkouts for genes to BDGP <i>in situ</i> and for alleles to FlyBase</li>
  <li>FIX - you can now access the lists and query history section of MyMine when not logged in.  This allows you to delete your temporary lists.</li>
  <li>FIX - GLEANR symbols are no longer low quality matches when using them to upload a list of <i>Drosophila</i> genes.</li>
  <li>FIX - template and list search pages no longer reset the search term when using the back button.</li>
</ul>
</p>

<h3>Template web services</h3>
<p>
<ul>
  <li>FIX - the 'Show link with constraint values highlighted in red' now highlights every value correctly</li>
  <li>FIX - various javascript errors in Internet Explorer 6.0 have been fixed that prevented the results count from being updated.</li>
  <li>NEW - results are returned with the correct file extension which makes it easier for programs to correctly open the file</li>
  <li>FIX - you can now login while on the template web services page</li>
  <li>FIX - template form correctly updates on error</li>
</ul>
</p>

<h3>Widgets</h3>
<p>
<ul>
  <li>NEW - extra columns added to the widget export</li>
  <li>FIX - enrichment widgets include more digits when displayed in scientific notation</li>
  <li>FIX - enrichment widgets no longer overlap when displayed</li>
  <li>NEW - statistical enrichment widgets now link to documentation on the method of calculation used</li>
</ul>
</p>

</div>

<div class="heading2">FlyMine version 13.0 - release notes (10th June 2008)</div>
<div class="body">
<p style="font-size: +1.1em;">
In Release 13.0 we have improved integration of data with old <i>Drosophila</i> gene identifiers from several sources, this information is now applied to a current FlyBase gene where possible.  We have also added allele and phenotype information from FlyBase and miRNA target predictions from miRBase.  Data from FlyBase is updated to the FB2008_05 release and other sources have been updated to the latest version.  User interface improvements focus on uploading and saving lists.  We have also launched a RESTful web service allowing queries to be run from scripts and remote sites.</p>

<h3>Data integration</h3>
<p>In this release we have made a major effort to update old identifiers for <i>Drosophila</i> genes that are provided by many data sources to match a current FlyBase identifier.  We have done this with a synonym lookup, for example when we download data from a particular source that references gene CG12802 which is no longer a current gene identifier but is a synonym (an old identifier) for FBgn0250818.  Where an old identifier did not match a single gene we did not load data for it.  As a result we have improved data integration, you should see more data attached to current genes and no longer see old genes with data from only one source.</p>

<h3>Data</h3>
<p>We have added <i>Drosophila</i> allele and phenotype data from FlyBase and miRNA target predictions from miRBase.  Data from FlyBase is updated to the FB2008_05 release and many other sources have been updated to the latest versions, new data are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>
<p>
<ul>
  <li>NEW - <i>Drosophila melanogaster</i> allele and phenoytype data from FlyBase.</li>
  <li>NEW - <i>Drosophila melanogaster</i> predicted miRNA targets from miRBase.</li>
  <li>UPDATE - three new RNAi screens added from DRSC.</li>
</ul>
</p>

<h3>Saving lists</h3>
<p>We have enhanced creation of lists from results tables and fixed some selection issues.</p>
<p>
<ul>
  <li>NEW - a 'Selection' bar below the results table tracks objects that have been selected.</li>
  <li>NEW - where the same object appears on multiple rows, selecting it once will select all rows even if on different pages.</li>
  <li>FIX - performance of list creation from results improved.</li>
  <li>FIX - selecting individual objects from multiple pages now works correctly.</li>
  <li>FIX - selecting a whole column then de-selecting some rows now works as expected.</li>
</ul>
</p>

<h3>Web Services</h3>
<p>We have begun to introduce RESTful web services to enable access to FlyMine data from scripts and other web sites.  We are working on adding client APIs and further features.</p>
<p>
<ul>
  <li>It is now possible to run an XML query via a simple URL and get results as XML, or tab separated formats.  See <a href="http://www.intermine.org/wiki/WebService">documentation</a> for more.</li>
  <li>All public template queries can be executed for any values and HTML results embedded in a remote web page.  So if you run a web site presenting <i>Drosophila</i> data you can fetch results live from FlyMine to view alongside your own.  Please contact us if you any questions about this.  See <a href="http://www.intermine.org/wiki/TemplateWebService">documentation</a> for more.</li>
</ul>
</p>

<h3>List upload</h3>
<p>
<ul>
  <li>NEW - organism name is now displayed for issues on list confirmation page.</li>
  <li>FIX - fixed problems uploading from a file on some versions of Safari.</li>
  <li>FIX - improved performance and fixed unresponsive script errors when clicking 'Add all' with very large numbers (hundreds) of issues on upload confirmation page.</li>
  <li>FIX - improved error messages for incorrect file names typed into file upload box.</li>
  <li>FIX - improved disabling of paste/browse options on list upload page.</li>
</ul>
</p>

<h3>User interface</h3>
<p>
<ul>
  <li>NEW - the List Analysis page menu bar has a 'Find' option to search for a particular identifier in the list.</li>
  <li>NEW - the Lists tab now has a 'Copy' option.  Select a list and type a new name to create a copy, or select multiple lists and they will be named automaticaly.  You can copy public lists as well your own lists.</li>
  <li>NEW - option to show/hide descriptions in the Template and List tabs.</li>
  <li>NEW - widgets on List Analysis pages now show a count of objects that have no data for the widget.</li>
  <li>FIX - linking to FlyBase with a list of genes from the List Analysis page now works with large lists.</li>
  <li>NEW - link to FlyExpress with a list of genes from List Analsyis page.</li>
  <li>NEW - added links to external databases in enrichment widgets.</li>
  <li>NEW - link to BDGP <i>in situ</i> data from gene report pages. </li>
</ul>
</p>

</div>

<div class="heading2">FlyMine version 12.0 - release notes (30th March 2008)</div>
<div class="body">
<p style="font-size: +1.1em;">
Release 12.0 of FlyMine adds genome annotation for the ten additional <i>Drosophila</i> species included in FlyBase release FB2008_02.  Homology data between the 12 <i>Drosophila</i> species from the 12 Genomes Consortium (2007) paper.  Data from BDGP <i>in situ</i> and genetic interactions from BioGRID are also added.  We have also added more widgets for analysing lists of genes and improved the style and layout of the list analysis page.</p>

<h3>Data</h3>
<p>Many existing data sources have been updated to the latest versions, new data sources are detailed below.  For a complete list of all data included in FlyMine see the 'Data' tab above.</p>
<ul>
  <li>NEW - <i>Drosophila</i> 12 genomes.  We have added genome annotation for the additional ten sequenced <i>Drosophila</i> genomes from FlyBase release FB2008_02.</li>
  <li>NEW - <i>Drosophila</i> 12 genomes homology - homology between 12 <i>Drosophila</i> genomes, data from the 12 Genomes Consortium (2007).  PubMed 17994087</li>
  <li>NEW - BDGP in situ expression patterns.</li>
  <li>NEW - Genetic Interaction data for <i>D. melanogaster, C. elegans</i> and <i>S. cerevisiae</i> from BioGRID.</li>
  <li>NEW - Probe sets from the Affymetrix GeneChip Drosophila Genome 1.0 Array.</li>
  <li>FIX - Some InParanoid cluster names have been fixed.</li>
  <li>FIX - ProteinDomain name and shortName were both showing the short name, now fixed.</li>
  <li>FIX - <i>Drosophila</i> UTRs and CDSs now have FlyBase as evidence.</li>
  <li>FIX - Evidence for REDfly TFBindingSites now fixed.</li>
</ul>

<h3>List Analysis</h3>
<p>The layout of list analysis pages has been improved.  New 'List Info' and 'Link Outs' sections are added to the top ot the page.  Widgets have had a style update, more can be displayed on the screen at once and a menu allows you to open/close them.  Loading times of the list analysis page has been greatly improved.  Details of new widgets are given below.</p>
<ul>
  <li>NEW - Orthologues widget - for a list of genes shows counts of orthologues in other organisms using data from InParanoid and homology between the 12 <i>Drosophila</i> genomes.</li>
  <li>NEW - BDGP in situ expression patterns widget. For a list of genes display the counts expressed or not expressed at the stage ranges tested.</li>
  <li>NEW - BDGP in situ term enrichment.  For a list of genes see if any ImaGO terms assigned by BDGP are statistically enriched.</li>
  <li>NEW - Genetic Interactions widget.  For a list of genes show genes that genetically interact and the count of genes from the list that they interact with.  Data from BioGRID.</li>
  <li>NEW - For a list of <i>Drosophila</i> genes export all directly to FlyBase's HitList page.</li>
  <li>NEW - Data from enrichment widgets can now be exported as tab or comma separated files.</li>
  <li>FIX - no longer get an exception after adding elements to an existing list then clicking 'view all results' on list analysis page.</li>
</ul>

<h3>Other fixes</h3>
<ul>
  <li>FIX - Column summary buttons and option to save a list now appear on all results tables.</li>
  <li>FIX - Exporting results has been made much faster.</li>
  <li>FIX - GFF3 export now only exports unique rows.</li>
  <li>FIX - Bug with exporting results in IE6 fixed.</li>
  <li>FIX - List upload now ignores duplicate identifiers even if the case doesn't match.</li>
  <li>FIX - Links to KEGG pathways on the KEGG web site have been corrected.</li>
  <li>FIX - Improved error messages when importing invalid query XML.</li>
  <li>FIX - Constraining a field in the QueryBuilder to 'has no value' sometimes didn't work, now fixed.</li>
</ul>
</div>


<div class="heading2">FlyMine version 11.0 - release notes (7th February 2008)</div>
<div class="body">
<p style="font-size: +1.1em;">
Release 11.0 of FlyMine includes an update of data from FlyBase, expression data related to Tiffin motifs has been added and many other data sources are updated.  We have also added new widgets and improved menus on results and list analysis pages.</p>

<h3>Data</h3>
<p>The <i>D. melanogaster</i> genome has been updated to release 5.5 from FlyBase.  Many other data sources have been updated to recent versions, see the Data tab for more details.</p>
<ul>
  <li>FIX - We now follow part-of relationships when computing transitive closure on the the Gene Ontology.  This means that queries that include parents/children of terms now follow part-of relationships as do the GO enrichment widgets.</li>
  <li>FIX - <i>D. melanogaster</i> UTRs and CDSs are now retrieved correctly from FlyBase</li>
  <li>FIX - Some minor changes have been made to the ProteinInteraction class - unused attributes have been removed, interacting regions is now filled in.</li>
  <li>UPDATE - InterPro domains for each protein are now retrieved from UniProt with details of the domains taken from InterPro.  We now include domains for more organisms.</li>
  <li>UPDATE - As of October 2007, the <i>Drosophila</i> DNAse I footprint (FlyReg) database has fully merged with the REDfly database. The FlyReg website will no longer be updated. Therefore, FlyMine now loads the FlyReg transcription factor binding sites directly from the REDfly database.</li>
  <li>UPDATE - All REDfly identifiers have been updated.</li>
  <li>UPDATE - <i>C. elegans</i> RNAi phenotype data from WormBase have been remodeled. Every phenotype now has a flag for 'Observed': true/false. Some phenotypes show a penetrance range.</li>
  <li>NEW - Five more <i>D. melanogaster</i> RNAi screens added from the DRSC.</li>
  <li>NEW - A subset of Tiffin motifs are now correlated with known expression patterns.</li>
</ul>

<h3>Orthologues</h3>
<p>
We have changed the way we present orthologue data from <a href="http://inparanoid.sbc.su.se/">InParanoid</a> to make it simpler and more obvious.  Orthologues and Paralogues now appear in a single Homologue class which has a type of 'orthologue' or 'inParalogue', we no longer use the terms 'main' and 'secondary'.  Please contact us if you have any problems with this.
</p>
<p>
When viewing a list of genes on a list analysis page there is now an orthologue conversion section in the top right.  Select an organism to view a results page of orthologues in that organism to export or create a new list.
</p>

<h3>Widgets</h3>
<p>
New widgets have been added to protein and genome feature list analysis pages.  Enrichment widgets now include the option to choose Benjamini-Hochberg or Bonferroni error correction.  Benjamini-Hochberg is the default and is less stringent.  Enrichment and table widgets can now be opened in a new window and have the option to select multiple categories (e.g. multiple GO terms) and display/export all data from those categories.
<ul>
  <li>NEW - A chromosome distribution widget is now shown on list analysis pages for any located genome feature (e.g. genes, exons, binding sites).</li>
  <li>NEW - Enrichment widgets added to protein list analysis page:
    <ul>
      <li>InterPro protein domains.</li>
      <li>GO terms (queried via corresponding genes).</li>
      <li>UniProt protein features.</li>
      <li>UniProt keywords.</li>
    </ul>
  </li>
  <li>NEW - Enrichment widget results can be filtered on maximum p-value.</li>
  <li>FIX - clicking on bars of chromosome distribution widget now gives correct results with multi-organism lists.</li>
  <li>FIX - links on GO enrichment now give results for <i>A. gambiae</i> and <i>A. mellifera</i></li>
</ul>
</p>
<h3>List upload</h3>
<ul>
  <li>NEW - 'Add all'/'Remove all' links added to each issue section on list upload page - making large lists easier to manage.</li>
  <li>NEW - Issues on the confirmation page now have links to open report pages in a new window.</li>
  <li>FIX - Confirmation page no longer resets if an invalid list name is entered.</li>
  <li>FIX - List name is no longer pre-filled with the last entered list name.</li>
</ul>

<h3>Other user interface changes</h3>
<ul>
  <li>NEW - Template query forms include an 'XML' link to export an XML representation of the query.</li>
  <li>NEW - Export, Create List and Add To List are now in a menu bar above results tables, combined with previous/next navigation.</li>
  <li>NEW - Open/closed state of sections on report and list analysis sections is now 'sticky' for your current session.</li>
  <li>FIX - Improvements to error checking and messages when importing XML queries/templates.</li>
  <li>FIX - Selections don't reset if an invalid name entered when performing list combination operations.</li>
  <li>FIX - Improved error handling on results page when selecting objects to save and list name.</li>
</ul>
</div>
<div class="heading2">FlyMine version 10.0 - release notes (4th December 2007)</div>
<div class="body">
<p style="font-size: +1.1em;">
Release 10.0 of FlyMine includes an update of data from FlyBase, a new <i>A. gambiae</i> gene expression data set and some user interface changes.
</p>
<h3>Data</h3>
<p>
The <i>D. melanogaster</i> genome has been updated to release 5.4 from FlyBase.  We have added a new microarray-based gene expression data set for the life cycle of <i>A. gambiae</i>.  'Life cycle transcriptome of the malaria mosquito Anopheles gambiae and comparison with the fruitfly Drosophila melanogaster.' (Koutsos et al 2007 - <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;dopt=Abstract&amp;list_uids=%2017563388">PubMed</a>).  
</p>
<p>
Other data changes:
<ul>
  <li>FIX - duplicate introns have been removed for <i>D. melanogaster</i> and <i>D. pseudoobscura</i>, FlyMine used to infer introns but they are now imported from FlyBase.</li>
  <li>FIX - introns are now created on the correct strand for <i>A. gambiae</i> and <i>A. mellifera</i>.</li>
  <li>FIX - orthologues for <i>A. gambiae</i>, <i>A. mellifera</i> and <i>A. aegypti</i> now reference genes as well as translations, they will now appear in results of more template queries.</li>
  <li>FIX - <i>D. pseudoobscura</i> 'Unknown' assembly groups are no longer created as chromosomes.</li>
  <li>FIX - the 'localised' category has been removed for fly-FISH expression data, the data now just shows whether a gene is expressed or not at a particular stage.</li>
</ul> 
</p>
<h3>List Conversion</h3>
<p>The List analysis page has a new 'Convert' section for instant conversion between common types (genes, proteins, transcripts and exons).  Counts of each type are displayed for a list - e.g. the numbers of proteins, transcripts and exons found for a list of genes.  Clicking on the type will bring up a results table from which you can export identifiers, sequence or create a new list.
</p>

<h3>External links</h3>
<p>We have added links to more external databases from report pages - e.g. to model organism databases and REDfly.  We now have a simple configuration system for links so if you have suggestions of other databases it would by useful to link to please let us know - support[at]flymine.org.
</p>

<h3>Tutorial and FAQ</h3>
<p>We have added a tutorial on creating and using lists in FlyMine, see the <a href="/help.shtml">help page</a>.  There is also a list of <a href="http://www.intermine.org/wiki/FlyMineFAQ">Frequently Asked Questions</a> linked to from the header.
</p>

<h3>User interface changes</h3>
<ul>
  <li>NEW - the queries run when clicking on graphs or tables in list analysis widgets now include more information.</li>
  <li>NEW - number of rows shown in results tables is remembered.</li>
  <li>FIX - performance improvement for uploading lists with more than 1000 elements.</li>
  <li>FIX - fixed problem where user was logged out or history was lost when returning to the home page.</li>
  <li>FIX - uploading a list from a file is fixed on Safari.</li>
  <li>FIX - it is now more obvious when long text entries have been truncated in results tables.</li>
  <li>FIX - fixed 'div has no properties' error in templates tab if user created two templates with the same title.</li>
</ul>
</div>


<div class="heading2">FlyMine version 9.0 - release notes (8th November 2007)</div>
<div class="body">
<p style="font-size: +1.1em;">
In release 9.0 we have updated the <i>D. melanogaster</i> genome to release 5.3 and the <i>A. gambiae</i> genome to release AgamP3.4.  We have added mRNA localisation data from Fly-FISH (<a href="http://www.utoronto.ca/krause/" target="_new">Krause lab</a>) and updated many other sources of data.
</p>

<h3>Genome Updates</h3>
<p>
The <i>D. melanogaster</i> genome has been updated to release 5.3 from FlyBase.  The <i>A. gambiae</i> genome has been updated to release AgamP3.4 from Ensembl.  Note that with this genome build the identifier scheme has changed - the primary identifiers for <i>A. gambiae</i> features now used in FlyMine are those from VectorBase.
</p>

<h3>Fly-FISH data</h3>
<p>
mRNA subcellular localisation data for <i>D. melanogaster</i> from Fly-FISH has been added.  There is a new widget on the gene list analysis page to display the numbers of genes in the list that are localised, not localised or not expressed. 
</p>

<h3>Other data changes</h3>
<p>
Many data sources have been updated to the latest available versions.  See the Data tab for more details.
</p>
<p>
<ul>
  <li>NEW - Four more <i>D. melanogaster</i> RNAi screens added from the DRSC.</li>
  <li>UPDATE - All 14 screens from the DRSC have a standardized results annotation, making it easier to query all of them simultaneously.</li>
  <li>FIX - first authors of publications are now set correctly.</li>
  <li>FIX - <i>D. melanogaster</i> Affymetrix probes no longer have genome locations - the locations were previously for an old genome assembly.</li>
  <li>FIX - <i>A. mellifera</i> genes and proteins are now linked together where possible, GO annotation is now attached to genes.</li>
  <li>FIX - evidence for proteins from UniProt is correctly set to Swiss-Prot or TrEmbl.</li>
</ul>
</p>

<h3>User interface changes</h3>
<p>
<ul>
  <li>NEW - spaces can be used as separators again in the list upload page - gene names with spaces should be quoted - e.g. "even skipped".</li>
  <li>NEW - widgets and template sections are now collapsed in list analysis page to improve page navigation.</li>
  <li>FIX - list upload from a file now works on Safari.</li>
  <li>FIX - FASTA export for Translation sequence is now fixed.</li>
  <li>FIX - links to VectorBase now correctly access gene pages.</li>
  <li>FIX - annoying 'Found 1 match(es) message removed from results page.</li>
  <li>FIX - correct template name/description displayed when clicking on 'Show in table' in a report page.</li>
  <li>FIX - some classes weren't listed in the class chooser in the QueryBuilder tab, now fixed.</li>
  <li>FIX - text in template form dropdowns should now always be displayed correctly.</li>
  <li>FIX - fixed bug with setting ascending/descending sort order in QueryBuilder.</li>
</ul>
</p>
</div>

<div class="heading2">FlyMine version 8.2 - release notes (17th October 2007)</div>
<div class="body">
<p style="font-size: +1.1em;">
In release 8.2 we have added new ways to link to FlyMine, optional organism constraints to many templates and changed the publication widget on the list analysis page to show statistical enrichment.
</p>

<h3>Link to FlyMine</h3>
<p>
In response to requests from other sites you can now link directly to FlyMine list analysis pages with any list of identifiers.  List analysis pages show graphs, tables and statistical enrichment for properties of a list - e.g. gene expression, protein domains, publications and pathways.  You can also link into FlyMine with, for example, a protein identifier and request to see the corresponding gene report page.  See <a href="/link.shtml">this document</a> to find out how to create links.
</p>

<h3>Publication Enrichment</h3>
<p>
The publication widget on the gene list analysis page now shows statistical enrichment of related publications.  This is a great way to find the most relevant publications for a given gene list and may help you interpret the list.  Gene/publication relationships are loaded from <a href="http://www.ncbi.nlm.nih.gov/sites/entrez" target="_new">PubMed</a>.
</p>

<h3>Optional organism constraints</h3>
<p>
Many templates now have an optional constraint on organism to make it easier to use these templates with gene symbols.
</p>

<h3>Other changes</h3>
<p>
<ul>
  <li>FIX - template form layouts have been improved.</li>
  <li>FIX - totals are now displayed correctly in the FlyAtlas widget on list analysis pages.</li>
  <li>FIX - wildcards now work in quick search again (e.g. you can search for 'zen*').</li>
  <li>FIX - Lists/Templates pages are now correctly reset if text is removed from filter box.</li>
  <li>FIX - logging out on some pages gave an error message, this is fixed.</li>
</ul>
</p>
</div>



<div class="heading2">FlyMine version 8.1 - release notes (11th September 2007)</div>
<div class="body">
<p style="font-size: +1.1em;">
In release 8.1 of FlyMine we have made some major changes to the layout of FlyMine.  This has been done in response to feedback from users and to make the most exciting parts of FlyMine more obvious.  We have also added public lists and a new help system.
</p>

<h3>New Layout</h3>
<p>
A new <strong>header</strong> provides tabs for easy access to the major areas of the site.  The <strong>Templates</strong> and <strong>Lists</strong> tabs link to new search pages that show all templates/lists and update as you type keyword search terms.  You can also restrict by data category and, if you are logged in, restrict to your favourites or own saved data.  The <strong>search bar</strong> in the header is a quick way to search identifiers and now also templates and lists.
</p>

<p>
The updated <strong>home page</strong> layout introduces the main concepts - Data categories, Templates, Lists and the QueryBuilder.  There is also space for us to highlight new or interesting templates and public lists.
</p>

<h3>Renaming</h3>
<p>
In response to user feedback we have renamed some elements of the FlyMine interface to make them more obvious.  <strong>'Bags'</strong> are now called <strong>'Lists'</strong>, <strong>'Aspects'</strong> are now <strong>'Data categories'</strong>.  We also refer now refer to <strong>Report pages</strong> instead of the computer speak: <strong>Object details pages</strong> and a <strong>List analysis page</strong> instead of a <strong>Bag details page</strong>.  Hopefully it won't cause confusion for too long.  
</p>

<h3>Public Lists</h3>
<p>
Another major development is the introduction of <strong>public lists</strong>.  These are lists that we create and make available for everyone to use, typically they will be derived from a publication.  In this release we have added public lists based on <a href="http://www.flyatlas.org" target="_new">FlyAtlas</a> expression in the adult fly tissues and <a href="http://www.flytf.org" target="_new">FlyTF</a> transcription factor lists.  Click on an example on the home page to see a list analysis page with graphs, tables, statistical enrichment and the results of templates run for the list.  We can add new public lists at any time, please contact us if you have a suggestion.
</p>

<h3>List Analysis Widgets</h3>
<p>
We have added two new <strong>'widgets'</strong> to the analysis page for a list of genes.  These show statistical enrichment of InterPro protein domains within a list of genes and the publications that reference the most genes from the list (data from PubMed).  We will continue to add to the list analysis page in every release, so let us know if you have any suggestions.
</p>

<h3>Help System</h3>
<p>
A new context sensitive help system has replaced the FlyMine manual.  On every page there is a '?' icon to the top right, click this to get help with the current page and find out what you could do next.  The help should now be briefer and more readable.
</p>

<h3>Other Changes</h3>
<br/>
<ul>
  <li><strong>Results page</strong>
    <ul>
      <li>FIX - column summaries sometimes gave misleading results when counting unique rows.</li>
      <li>FIX - the total number of rows and unique rows in summaries now update correctly in Safari.</li>
    </ul>
  </li>
  <li><strong>MyMine</strong>
    <ul>
      <li>FIX - colons are now allowed in list/template names.</li>
      <li>FIX - occasional issues with creation dates of templates/lists are fixed.</li>
    </ul>
  </li>
  <li><strong>Report page</strong>
    <ul>
      <li>NEW - the bottom of each report page shows any lists (public or saved) that contain the object.</li>
      <li>NEW - template descriptions are now included when inline tables are open.</li>
    </ul>
  </li>  
  <li><strong>QueryBuilder</strong>
    <ul>
      <li>FIX - some issues with the view list and sort order on Internet Explorer are fixed.</li>
      <li>FIX - when using LOOKUP constraints can now convert wildcards that match an incorrect type - e.g. convert EVE_DROME* (a protein identifier) to a gene.</li>
      <li>FIX - LOOKUP constraints and list upload now accept names that contain spaces - e.g. 'even skipped' for the eve gene.</li>
      <li>FIX - sort order is now remembered for saved queries, doesn't get reset when removing elements from view list.</li>
    </ul>
  </li>
</ul>


</div>


<div class="heading2">FlyMine version 8.0 - release notes (27th June 2007)</div>
<div class="body">
<p style="">
      In release 8.0 we have updated the data model for protein interactions, protein structure and orthologues to make them easier and faster to query.  We also include three new data sets: Tiffin (predicted regulatory motifs), anoEST (A. gambiae EST clusters) and predicted 3-D protein structures for A. gambiae domains. A new type of constraint is used in template queries which now allows you to enter any identifier or symbol.
    </p>
<ul>
      <li>
        Data<br />
        <ul>
    <li>NEW - D. melanogaster predicted regulatory motifs and functional sites (motif instances) from the <a href="http://servlet.sanger.ac.uk/tiffin/">Tiffin database</a>. Note these have been mapped to the release 5.0 genome sequence.<br />
</li>
    <li>NEW - A. gambiae EST and EST clusters from the <a href="http://agambiae.vectorbase.org/SequenceData/EST" target="_new">Imperial College London Centre for Bioinformatics</a>.<br />
</li>
    <li>NEW - A. gambiae protein domain 3-D structure predictions from <a href="http://www-cryst.bioc.cam.ac.uk/~kenji/NEW/index.htm" target="_new">Kenji Mizuguchi</a>.  This data provides structure predictions for regions of A. gambiae proteins that correspond to a Pfam domain.    <br />
</li>
    <li>NEW - H. sapiens and M. musculus protein information included from UniProt, we already include InParanoid orthologues to these organisms.<br />
</li>
       </ul>
      </li>

      <li>
        User Interface<br />
        <ul>
    <li>NEW - LOOKUP constraints.  Most template queries are now updated to use a new type of constraint.  Now you don't need to enter a specific type of identifier - for example for the D. melanogaster gene zen you could enter 'zen', 'CG1046', 'FBgn0004053' or 'zerknullt' and get the same result.  You could even enter a protein and or transcript identifier and it would find the gene.<br />
</li>    
       </ul>
      </li>
    </ul>
</div>
</div> 

  <div style="clear:" class="box">
<div class="heading2">FlyMine version 7.1 - release notes (10th May 2007)</div>
<div class="body">
<p style="">
      Release 7.1 includes a number of new user interface features.  Results of queries can be ordered by a selected column and results tables have a new summary button which brings up statistics on the values in that column.  Many template queries now have more descriptive column titles in results tables.
    </p>
<ul>
      <li>
        User Interface<br />
        <ul>
    <li>NEW - Column summaries in results tables.  Each column of a results table now has a summary icon, clicking this will bring up a box with more information about data in the column.  For numerical data it will show the minimum value, maximum value, mean and standard deviation.  For text it will display the number of unique values and the most commonly occurring values with their frequency.<br />
</li>
    <li>NEW - Sorting query results.  The QueryBuilder allows you to select an element from the output to sort results by.  A sort button lets you choose ascending or descending order, for example to display results with the highest confidence score or most recent publication first.<br />
</li>
    <li>NEW - Results column titles.  Many template queries are now configured to have more descriptive column headings in results tables.  The full path can be seen by hovering the mouse pointer over the description.<br />
</li>
    <li>NEW - Chromosome distribution viewer. The gene details page for <i>D. melanogaster</i> and <i>A. gambiae</i> now includes a chromosome distribution viewer.  This shows how many genes from the bag are found on each chromosome, click on a bar to see a list of the genes.  The graph also shows an expected number of genes for each chromosome based on the distribution of all genes between chromosomes and the size of the bag.<br />
</li>
    <li>NEW - Accurate counts on the results page.  The results page used to show only an approximate number of rows returned from a query (unless the 'Last' link was clicked).  The estimate is now updated to give an accurate 'Total rows' figure once it has been calculated. <br />
</li>
    <li>UPDATE - The trail (e.g. Query -&gt; Results -&gt; Gene -&gt; Protein) is now more complete to allow easy navigation back to recently viewed queries, results, object details or bag pages.<br />
</li>
    <li>FIX - Performance has been improved when saving and viewing large bags of objects.<br />
</li>
    <li>FIX - Renaming bags now works correctly. <br />
</li>
    <li>FIX - Missing export options from results pages have been fixed.  Genome features can be exported as FASTA or GFF3, protein interactions can be exported in Cytoscape SIF format.<br />
</li>
    <li>FIX - Sequences from Translation objects can now be exported as FASTA.<br />
</li>
    <li>FIX - Some minor issues with display on Internet Explorer have been fixed..<br />
</li>
        </ul>
      </li>

      <li>
        Tools<br />
          <ul>
    <li>FIX - TFModules from REDfly are now shown in GBrowse and have a GBrowse image on their details pages.<br />
</li>
         </ul>
      </li>

      <li>
        Known issues<br />
        <ul>
    <li>There are currently no known problems with release 7.1.<br />
</li>
        </ul>
      </li>
    </ul>
</div>
</div>

  <div style="clear:" class="box">
<div class="heading2">FlyMine version 7.0 - release notes (3rd April 2007)</div>
<div class="body">
<p style="">
      Release 7.0 updates the <i>D. melanogaster</i> genome to 5.1 annotation and other genome annotation sources have been re-mapped.  GO enrichment and KEGG pathway viewers have been added to the gene Bag Details page.
    </p>
<ul>
      <li>
        Data<br />
        <ul>
    <li>UPDATE - The <i>D. melanogaster</i> genome has been updated to annotation version 5.1.  Data from DrosDel, FlyReg, REDfly and the microarray tiling path have been re-mapped using <a href="http://genome.ucsc.edu/cgi-bin/hgLiftOver" target="_new">USCS LiftOver</a>.<br />
</li>
    <li> NEW - UniProt keywords (e.g. Acetylation, Sulfate transport) and protein features (e.g. HELIX, DNA_BIND) have been added.<br />
</li>
    <li>NEW - KEGG pathway information added for <i>D. melanogaster</i>.<br />
</li>
    <li>NEW - FlyAtlas now has data for three more tissues - larval fat body, larval tubule and male accessory gland.<br />
</li>
    <li>UPDATE - InParanoid orthologues have been updated to a release from January 2007.<br />
</li>
    <li>UPDATE - Four more D. melanogaster RNAi screens added from the DRSC the RNAi aspect.<br />
</li>
    <li>UPDATE GO annotation, protein-protein interactions and UniProt protein data are all updated to recent releases.<br />
</li>
    <li>FIX -  Missing protein structure data has been added, protein structures can now be viewed with JMol again.<br />
</li>
    <li>FIX - Missing protein interaction detection method has been replaced.<br />
</li>
    <li>FIX - Missing INDAC oligo sequences added.<br />
</li>
       </ul>
      </li>

      <li>
        User Interface<br />
        <ul>
    <li>FIX -Drosophila gene names now work in the quick search box, a number of minor problems with quick search have been fixed.<br />
</li>
    <li>NEW - The trail (e.g. Query -&gt; Results -&gt; Gene) has been improved for easy navigation between queries, results, and details pages.<br />
</li>
    <li>NEW - GO enrichment widget on the gene Bag Details page.  For the genes in the bag this lists the number of genes with a particular GO term and a p-value which is the probability that this number of genes were annotated with the GO term by chance, given the abundance of the GO term in a reference population.<br />
</li>
    <li>NEW - KEGG pathway widget on the gene Bag Details page.  This shows the number of genes in the bag that are associated with a particular KEGG pathway, links give the list of genes and more information about the pathway.<br />
</li>
    <li>UPDATE - The constraint editor pane of the QueryBuilder has been made clearer.<br />
</li>
    <li>NEW - import query from XML link added to the FlyMine home page.<br />
</li>
       </ul>
      </li>
    </ul>
</div>
</div>

  <div style="clear:" class="box">
<div class="heading2">FlyMine version 6.1 - release notes (15th February 2007)</div>
<div class="body">
<p style="">
FlyMine 6.1 contains a major overhaul of the way bags are handled.  Bags can now only contain actual objects rather than identifiers or symbols.  This means that any object in a bag has already been found in FlyMine which should reduce confusion.  A sophisticated bag upload system has been added to aid in creating bags from external lists of identifiers.
    </p>
<p style="">
PLEASE NOTE - most saved user content is automatically upgraded between FlyMine releases.  In this case it was not possible to port some types of bags.  These are still available in the 6.0 archive, please contact support [at] flymine.org if you have any queries about transferring bags.
    </p>
<p style="">
Bags also now have a type (class) assigned to them - for example Gene, Protein, GOTerm.  This means that when editing a constraint in a template query only bags of the correct type will be listed - so if the template requires you to enter a Gene identifier the bags dropdown will list any Gene bags in your profile.    The same is true when creating/editing a query in the QueryBuilder, just add a constraint on the identifier, name, etc of a class and you will see available bags.
    </p>
<p style="">
The 'Bags' page in MyMine (select 'Bags' or 'MyMine' from the top menu bar) now allows you to paste in a list of identifiers and select a type for the new bag.  The input can be a mixture of different identifier types, for example if you wish to create a bag of <i>Drosophila</i> genes if can be a mixture of CGxx, FBgnxx and symbols.  In the case where an object can't be found to that matches a particular input identifier, FlyMine will attempt to help.  For example if the input list contains a UniProt protein identifier, but you choose to make a gene bag, the website will attempt to find a related gene.  Any matches found in this way will be reported for you to choose which are added to your bag.
    </p>
<p style="">
As an example, when creating a Gene bag from these identifiers:  zen CG2328 FBgn0015379 Q8IML9_DROME unknown_name FlyMine will find a gene for each of the first three identifiers and find the gene for the Q8IML9_DROME protein.  The "unknown_name" will be reported as not found.
    </p>
<p style="">
Also new are bag details pages.  These are accessible for any of your saved bags in the 'Bags' tab of MyMine.  They have a similar layout to object details pages but run templates for all objects in your page.  On the page for gene bags is the first of many 'widgets' we plan to add for a viewing and analysis of data in bags.  Currently there a widget that graphs the genes from a bag that are over/under expressed in different tissues according to the FlyAtlas data set (www.flyatlas.org).  Note that clicking on any of the bars in this graph allows you to create a new bag of genes in that category.  More functionality will be available on these pages in release 7.0/
    </p>
</div>
</div>


  <div style="clear:" class="box">
<div class="heading2">FlyMine version 6.0 - release notes (24th November 2006)</div>
<div class="body">
<p style="">FlyMine 6.0 adds data from seven <i>Dropsophila</i> RNAi screens from the DRSC,
          microarray-based gene expression data from FlyAtlas and Transcriptional cis-regulatory modules 
          (CRMs) from REDfly.  Template queries have a new, succinct naming scheme and identifiers/names
          in results tables are now links to object details pages.  Other data sources have been
          updated/added and there are numerous interface improvements.</p>
<p style="">
NOTE - from release 6.0 FlyMine does not support saved bags of objects.  User's object bags have not been ported to release 6.0 but can still be retrieved from the release 5.0 archive.  If any user wishes to use information from old object bags we recommend exporting from 5.0 as a list of appropriate identifiers and uploading to release 6.0.  Please contact us if you require any help with this.   All other saved information has been transferred to release 6.0.
    </p>
<p style="">
   An upcoming release of FlyMine will introduce a new system for uploading and managing bags of data.
    </p>
<ul>
      <li>
        Data<br />
        <ul>
    <li>NEW - Microarray-based gene expression data for <i>D. melanogaster</i> from FlyAtlas.<br />
</li>
    <li>NEW - High-throughput cell-based RNAi screens from the <i>Drosophila</i> RNAi Screening Center.<br />
</li>
    <li>NEW - Transcriptional cis-regulatory modules (CRMs) for <i>D. melanogaster</i> from REDfly.<br />
</li>
    <li>NEW - Probe sets from the Affymetrix GeneChip Drosophila Genome 2.0 Array.<br />
</li>
    <li>NEW - Syntenic regions between <i>D. melanogaster</i> and <i>D. pseudoobscura.</i>
<br />
</li>
    <li>NEW - Orthologues and GO annotation for <i>S. pombe</i> and <i>P. falciparum.</i>
<br />
</li>
    <li>NEW - Anatomy ontology for <i>Drosophila</i>.<br />
</li>
    <li>UPDATE - <i>D. melanogaster</i> genome annotation updated to version 4.3.<br />
</li>
    <li>UPDATE - UniProt protein data is now at version 8.9.<br />
</li>
    <li>UPDATE - Worm RNAi data from WormBase 30th September 2006.<br />
</li>
    <li>UPDATE - <i>D. melanogaster</i> BACs are now loaded.<br />
</li>
    <li>UPDATE - Protein fragments from UniProt are now identified by an isFragment true/false attribute.<br />
</li>
    <li>UPDATE - Proteins now have a length (in amino acids) and molecularWeight (in Daltons).<br />
</li>
    <li>FIX - Protein interaction confidence scores restored.<br />
</li>
    <li>FIX - PCRProducts now reference the TilingPathSpan they create.<br />
</li>
    <li>FIX - Some unpublished microarray experiments have been removed.<br />
</li>
       </ul>
      </li>

      <li>
        User Interface<br />
        <ul>
    <li>NEW - Names and identifiers in results tables now link to object details pages.<br />
</li>
    <li>NEW - Template names have been changed so that they now have a short name and a longer description. The short name is the main name and is displayed on the aspect pages, template search results and the object details pages. This shorter name has the format 'query starting point(s)' --&gt; 'query output' and allows easier scanning of templates to find the one required. The longer description can be viewed when the template form is accessed.<br />
</li>
    <li>NEW - When running a template query the title and description is displayed above the results.<br />
</li>
    <li>NEW - The trail now includes a 'Query' link back to the template or Query Builder, i.e. now 'Query -&gt; Results'.<br />
</li>
    <li>UPDATE - It is no longer possible to create bags of objects from results or upload.<br />
</li>
    <li>UPDATE - Chromosomal locations are displayed in a more compact format (e.g. 2R:1598168-1676472).<br />
</li>
    <li>UPDATE - 'History' has been renamed to 'My Mine' and the layout improved.<br />
        <ul>
          <li>improved query and template XML import/export options.<br />
</li>
          <li>added a 'Change Password' tab.<br />
</li>
          <li>bag listing and upload pages have been combined.<br />
</li>
        </ul>
</li>
    <li>FIX - Searches in bags of strings are no longer case sensitive.<br />
</li>
    <li>FIX - After running a template query 'Current Query' will now return to the template form instead of the Query Builder page.<br />
</li>
    <li>FIX - To address a performance issue bag size is limited to 100 entries unless you are logged in (when the limit is much higher). <br />
</li>
       </ul>
      </li>

      <li>
        Tools<br />
        <ul>
          <li>There are no changes to tools available in release 6.0.<br />
</li>
        </ul>
      </li>

      <li>
        Known issues<br />
        <ul>
          <li>There are currently no known issues with data in release 6.0.<br />
</li>
        </ul>
      </li>
    </ul>
</div>
</div>


  <div style="clear:" class="box">
<div class="heading2">FlyMine version 5.0 - release notes (1st June 2006)</div>
<div class="body">
<p style="">FlyMine 5.0 adds genome annotation for <i>D. pseudoobscura</i> (release 2.0 from FlyBase) and <i>A. mellifera</i> (release 2.0 from Ensembl) with InParanoid orthologues to the other genomes in FlyMine.  Also added are InParanoid orthologues to many other species and GO annotation for some of them.  Interface improvements include the ability for users to tag favourite templates and to change their password.</p>
<ul>
      <li>
        Data<br />
        <ul>
          <li>NEW: Genome annotation for <i>D. pseudoobscura</i> version 2.0 from FlyBase.<br />
</li>
          <li>NEW: Genome annotation for <i>A. Mellifera</i> release 2.0 from Ensembl version 37.<br />
</li>
          <li>NEW: InParanoid orthologues from <i>D. melanogaster, D. pseudoobscura, A. gambiae, C. elegans</i> and <i>A. mellifera</i> to many other organisms.<br />
</li>
          <li>NEW: GO annotation (accessible via orthologues) for <i>M. musculus, R. ratticus, S. cerevisiae</i> and <i>A. mellifera</i>.<br />
</li>
          <li>NEW: Protein interaction and UniProt data for S. cerevisiae.<br />
</li>
          <li>NEW: IntergenicRegion class - objects to describe intergenic regions on chromosomes, sequence and overlapping features available. Genes now have references to upstream and downstream intergenic regions.<br />
</li>
          <li>UPDATE: Genome annotation for <i>A. gambiae</i> updated to build AgamP3 from Ensembl version 37.<br />
</li>
          <li>FIX: INDAC microarray oligos now have sequences.<br />
</li>
          <li>FIX: OMIM disease names are no longer duplicated, new templates added to the Disease aspect.<br />
</li>
          <li>FIX: mRNA sequences no longer include intron sequences.<br />
</li>
          <li>FIX: Protein Interactions - some interactions referenced isoform ids so data didn't merge correctly with other protein information. Now have correct UniProt accession.<br />
</li>
          <li>FIX: Protein interactions are now correctly assigned to Formstecher et al, 2005 - was fixed in IntAct.<br />
</li>
      </ul> 
      </li>
      <li>
        User Interface<br />
        <ul>
          <li>NEW: Favourite templates - logged in users can see a * beside template query names to select then as favourites. They are listed in a new 'Favourites' tab on the History page.<br />
</li>
          <li>NEW: Change password - users can now choose there own password when creating an account. Passwords can be changed when logged in by clicking on the username in the top menu bar.<br />
</li>
          <li>FIX: Missing class descriptions completed.<br />
</li>
          <li>NEW: Boolean fields are now constrained by a true/false dropdown.<br />
</li>
      </ul> 
      </li>
      <li>
        Tools<br />
        <ul>
          <li>FIX: GBrowse images on object details pages are now oriented correctly for features on reverse strand. Also now show absolute coordinates.<br />
</li>
      </ul> 
      </li>
      <li>
        Known Issues<br />
        <ul>
          <li>Due to a change in format of XML from IntAct there are no confidence scores for protein interaction experiments, we will release a fixed version as soon as possible.<br />
</li>
          <li>Gene.goAnnotation collection not filled in for A. Gambiae genes, need to use Gene.annotations instead.<br />
</li>
        </ul> 
      </li>
    </ul>
</div>
</div>
  <div style="clear:" class="box">
<div class="heading2">FlyMine version 4.0 - release notes (13th March 2006)</div>
<div class="body">
<p style="">Major changes to layout and content of summary pages. These pages can now display the results of template queries inline to give immediate access to data. The www.flymine.org home page now shows aspect icons and query options. Many other interface improvements. Addition of Arbeitman et al, 2002 Drosophila time course microarray data and human disease gene matches from Homophila.</p>
<ul>
      <li>
        Data<br />
        <ul>
          <li>NEW: microarray data from "Gene Expression During the Life Cycle of <i>Drosophila melanogaster</i>", Arbeitman et al 2002.<br />
</li>
          <li>NEW: added data human disease gene matches from Homophila (http://superfly.ucsd.edu/homophila/).<br />
</li>
          <li>NEW: more protein interaction experiments from IntAct - see Protein Interaction aspect.<br />
</li>
          <li>NEW: Gene now has an "allGoAnnotation" collection that contains the term assigned to a gene plus all the parents of the term. Using this collection to find genes with a particular term will find all genes with that term *and* any with a more specific term.<br />
</li>
          <li>ENHANCEMENT: MRNA now references threePrimeUTR and fivePrimeUTR.<br />
</li>
          <li>ENHANCEMENT: <i>D. melanogaster</i> translation symbols now set from FlyBase.<br />
</li>
          <li>ENHANCEMENT: <i>D. melanogaster</i> chromosome bands now loaded from FlyBase.<br />
</li>
          <li>UPDATE: <i>D. melanogaster</i> genome annotation updated from 4.2 to 4.2.1.<br />
</li>
          <li>UPDATE: GO annotation and UniProt proteins updated.<br />
</li>
          <li>FIX: <i>D. melanogaster</i> regulatory regions from FlyBase now have chromosome location.<br />
</li>
          <li>FIX: <i>D. melanogaster</i> insertion sites and point mutations from FlyBase reference genes correctly.<br />
</li>
          <li>FIX: <i>A. gambiae</i> genes from Ensembl and InParanoid orthologues now merged correctly.<br />
</li>
        </ul>
      </li>

      <li>
        Tools<br />
        <ul>
          <li>NEW: a graph of interacting proteins on protein summary pages.<br />
</li>
          <li>NEW: graphs of <i>D. melanogaster</i> time course expression on gene details pages (Arbeitman et al, 2002).<br />
</li>
          <li>NEW: FASTA retrieval icon to access sequence of all features with sequence.<br />
</li>
          <li>FIX: insertion sites now displayed correctly on GBrowse.<br />
</li>
        </ul>
      </li>

      <li>
        User Interface<br />
        <ul>
          <li>NEW: Template queries are now viewable inline on object summary pages.<br />
</li>
          <li>NEW: bags saved in history can now be subtracted from one another.<br />
</li>
          <li>NEW: Query constraint logic can now be manipulated within the query builder - can constrain A AND B/A OR B.<br />
</li>
          <li>NEW: Added Run query link to run a query directly from the History page.<br />
</li>
          <li>NEW: Template queries can now include a choice of several different output formats.<br />
</li>
          <li>ENHANCEMENT: Template search terms are highlighted in results lists.<br />
</li>
          <li>ENHANCEMENT: Template building integrated into query builder interface.<br />
</li>
        </ul>
      </li>

      <li>
        Known issues<br />
        <ul>
          <li>DATA: Protein interactions from Formstecher et al, 2005 (PubMed: 15710747) are incorrectly assigned to Rain et al, 2001 (PubMed: 11196647) which in IntAct is a secondary reference.<br />
</li>
          <li>DATA: OMIM disease names contain the same text repeated multiple times.<br />
</li>
        </ul>
      </li>

    </ul>
</div>
</div>

  <div style="clear:" class="box">
<div class="heading2">FlyMine version 3.1 - release notes (7nd November 2005)</div>
<div class="body">
<ul>
      <li>
        Data - fixes<br />
        <ul>
          <li>FIX: <i>D. melanogaster</i> genes from FlyBase and UniProt are now merged correctly.<br />
</li>
          <li>FIX: Missing orthologues from <i>D. melanogaster</i> to <i>A. gambiae</i> are now loaded.<br />
</li>
        </ul>
      </li>
    </ul>
</div>
</div>

  <div style="clear:" class="box">
<div class="heading2">FlyMine version 3.0 - release notes (1st September 2005)</div>
<div class="body">
<ul>
      <li>
        Data - new and updated<br />
        <ul>
          <li>UPDATE: FlyBase <i>D. melanogaster</i> genome annotation updated to release 4.2.<br />
</li>
          <li>NEW: Microarray gene expression data - currently five experiments run at the FlyChip facility are available.<br />
</li>
          <li>NEW: Protein domain and family information from InterPro.<br />
</li>
          <li>NEW: DrosDel deletion data added.<br />
</li>
          <li>NEW: <i>D. melanogaster</i> genes now have full name and wild type function set.<br />
</li>
          <li>UPDATE: GO annotation changes:<br />
            <ul>
              <li>
<i>D. melanogaster</i> and <i>C. elegans</i> gene associations are now taken from FlyBase and WormBase respectively.<br />
</li>
              <li>Gene now has a collection of GOAnnotations, an object which contains the GO term id, name, evidence code and text from 'with' reference if present. If 'with' is a Gene or Protein a reference to the object is added.<br />
</li>
              <li>GOTerm entries now have a description, namespace ('biological_process', 'cellular_component' or 'molecular_function') and a collection of synonyms.<br />
</li>
            </ul>
          </li>
        </ul>
        NOTE - this list does not include all updates to the versions of existing data sources - for a full list of versions see the Datasources section on the <a href="/">FlyMine home page</a>.<br />
        <br />
      </li>
      <li>
        Data - fixes<br />
         <ul>
          <li>FIX: Gene.name is now Gene.symbol, added Gene.name which has a longer text name.<br />
</li>
          <li>FIX: Transcript.exonCount is now set for all Transcripts.<br />
</li>
          <li>FIX: Only RepeatRegions predicted by TRF are retrieved for <i>A. gambiae</i> from Ensembl.<br />
</li>
          <li>FIX: There are no longer links to UniProt with BDGP identifiers.<br />
</li>
          <li>FIX: Incorrect start and end fields removed from RepeatRegion.<br />
</li>
          <li>FIX: Removed duplicate synonyms for some Proteins.<br />
</li>
        </ul>
      </li>

      <li>
        Data - known issues<br />
         <ul>
           <li>Five genes are annotated incorrectly in the FlyBase release 4.2 annotation. These are eys, Dimp, cp309, Nev3 and kay. It is a known problem with FlyBase and a fix is progress. In FlyMine the symbols(names) are modified so that they are unique and indicate that there is a problem. Examples are: Dimp-duplicate-symbol-1, Dimp-duplicate-symbol-2 instead of Dimp.<br />
</li>
           <li>The latest version of InterPro (11.0) is not yet available as a mysql dump so we still have version 10.0.<br />
</li>
           <li>INDAC microarray oligos were designed to FlyBase version 4.0 transcripts. This means a number now map to old transcripts. These oligos can be queried in FlyMine, however the transcript they are designed to will not have merged with a Gene.<br />
</li>
           <li>A number of proteins from Uniprot have not merged with FlyBase genes.  This means that for certain genes there are two objects - one from Uniprot with no other information.<br />
</li>
           <li>Orthologues for <i>A. gambiae</i> did not load.<br />
</li>
        </ul>
      </li>

      <li>
        Interface - new features<br />
        <ul>
          <li>NEW: The start page of FlyMine now has links to a series of 'aspects' which are convenient starting points for the different types of data in FlyMine - such as genomics, protein interactions and gene expression. Each aspect lists relevant data sets in FlyMine, template queries and suggested starting points for queries.<br />
</li>
          <li>NEW: Template search box added to the front page.<br />
</li>
          <li>NEW: User templates are now listed on the user's History page.<br />
</li>
          <li>NEW: Saved bags and saved queries can be renamed from the History page.<br />
</li>
          <li>NEW: Executed queries are now automatically saved to a 'session history'. The user can permanently save a query from the session history from the History page.<br />
</li>
          <li>NEW: Saved queries and queries in the session history are now displayed with the date and time at which they were created.<br />
</li>
          <li>NEW: Saved queries and queries in the session history are now displayed with their starting point and a summary of the fields selected for output.<br />
</li>
          <li>NEW: You can now edit an existing constraint in the query builder by clicking the blue edit icon next to the constraint.<br />
</li>
        </ul>
      </li>

      <li>
        Interface - fixes<br />
        <ul>
          <li>FIX: History page is now viewable to anonymous users.<br />
</li>
          <li>FIX: Characters from the Greek alphabet within identifier names should now display correctly.<br />
</li>
          <li>FIX: Users can now run separate queries in separate browser windows and view results simultaneously.<br />
</li>
          <li>FIX: Exported results on Windows Internet Explorer now download with the correct file name.<br />
</li>
        </ul>
      </li>
    </ul>
</div>
</div>
  <div style="clear:" class="box">
<div class="heading2">FlyMine version 2.1 - release notes (1st June 2005)</div>
<div class="body">
<ul>
      <li>
        Data - new sources<br />
        <ul>
          <li>Added <i>D. melanogaster</i> whole genome amplimer tiling path (Genetics Department, Cambridge, UK)<br />
</li>
        </ul>
      </li>

      <li>
        Data - fixes<br />
         <ul>
          <li>Fixed an off by one error with chromosome locations of FlyReg binding sites<br />
</li>
          <li>GBrowse images now appear on Transcript details page<br />
</li>
          <li>Removed duplicate GO (Gene Ontology) Ontology object<br />
</li>
          <li>Sequences are no longer a mix of upper and lower case - now all lower case<br />
</li>
          <li>Added CDS and Translation objects for each <i>A. gambiae</i> Transcript<br />
</li>
          <li>INDAC oligos - Transcripts now correctly merged with other <i>D. melanogaster</i> annotation<br />
</li>
          <li>All genome annotation features now have a chromosomeLocation field<br />
</li>
        </ul>
      </li>

      <li>
        Data - known issues<br />
         <ul>
          <li>Information from FlyBase missing for some <i>D. melanogaster</i> genes due to problems reading FlyBase gff3 (fixed in FlyBase release 4.1)<br />
</li>
          <li>INDAC, FlyReg and Genetics Dept (Cambridge) InfoSource objects are duplicated<br />
</li>
        </ul>
      </li>

      <li>
        Interface - new features<br />
        <ul>
          <li>Template queries now work with bags - an attribute can be constrained to be in/not-in a bag<br />
</li>
          <li>Template search facility added. Users can search public and private templates<br />
</li>
          <li>All string matches in FlyMine queries are now case insensitive<br />
</li>
          <li>Performance improvements while paging through results/collection tables<br />
</li>
          <li>Collection/bag table pages are now bookmarkable<br />
</li>
          <li>Keywords can be attached to template queries<br />
</li>
        </ul>
      </li>

      <li>
        Interface - fixes<br />
        <ul>
          <li>Users with javascript disabled can now run queries<br />
</li>
          <li>Fixed bug that stopped the user from creating certain loop-queries<br />
</li> 
        </ul>
      </li>
    </ul>
</div>
</div>
</div>
</div>
